### ISSUE TITLE
ArtifactHttpClient.request occasionally fails with "Unexpected token '<', "<!DOCTYPE "... is not valid JSON" error

### ISSUE THREAD

**Issue by contrast-jproberts** [actions/toolkit#1888]
**Describe the bug**
When using the `actions/download-artifact` and `actions/upload-artifact` actions this morning, I saw multiple jobs fail with `Failed to CreateArtifact: Unexpected token '<', "<!DOCTYPE "... is not valid JSON` on upload. I also saw some jobs fail with the same error message when attempting to GetSignedArtifactURL while download. Most jobs succeeded when I reran the failed jobs, but 1 failure persisted. Fortunately, it gave a stack trace with some more information:

```
/home/runner/work/_actions/actions/download-artifact/eaceaf801fd36c7dee90939fad912460b18a1ffe/dist/index.js:2889
                throw new Error(`Failed to ${method}: ${error.message}`);
                      ^

Error: Failed to GetSignedArtifactURL: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
    at ArtifactHttpClient.<anonymous> (/home/runner/work/_actions/actions/download-artifact/eaceaf801fd36c7dee90939fad912460b18a1ffe/dist/index.js:2889:23)
    at Generator.throw (<anonymous>)
    at rejected (/home/runner/work/_actions/actions/download-artifact/eaceaf801fd36c7dee90939fad912460b18a1ffe/dist/index.js:2841:65)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

I can't share the full log output here because the repository is private, but I've saved the output to answer followup questions if needed.

**To Reproduce**
Unfortunately, I haven't been able to consistently reproduce this error. Reruns often fix the issue.

If the issue happens again today, I'll rerun with debug logs which should give us more information.

**Expected behavior**
The action handles whatever error is happening, or gives a more meaningful error message to the user with suggested next steps.

**Additional context**
This could be the root cause for https://github.com/actions/upload-artifact/issues/543 and https://github.com/actions/download-artifact/issues/315 .

**Comment by contrast-jproberts** [comment]
I was able to reproduce with debug logs!

```
##[debug][Response] - 502
##[debug]Headers: {
##[debug]  "cache-control": "no-cache",
##[debug]  "content-type": "text/html; charset=utf-8",
##[debug]  "strict-transport-security": "max-age=31536000",
##[debug]  "x-content-type-options": "nosniff",
##[debug]  "x-frame-options": "deny",
##[debug]  "x-xss-protection": "0",
##[debug]  "content-security-policy": "default-src 'none'; style-src 'unsafe-inline'; img-src data:;",
##[debug]  "connection": "close"
##[debug]}
##[debug]Raw Body: <!DOCTYPE html>
##[debug]<!--
##[debug]
##[debug]Hello future GitHubber! I bet you're here to remove those nasty inline styles,
##[debug]DRY up these templates and make 'em nice and re-usable, right?
##[debug]
##[debug]Please, don't. https://github.com/styleguide/templates/2.0
##[debug]
##[debug]-->
##[debug]<html>
##[debug]  <head>
##[debug]    <title>Unicorn! &middot; GitHub</title>
##[debug]    <style type="text/css" media="screen">
##[debug]      body {
##[debug]        background-color: #f1f1f1;
##[debug]        margin: 0;
##[debug]        font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
##[debug]      }
##[debug]
##[debug]      .container { margin: 50px auto 40px auto; width: 600px; text-align: center; }
##[debug]
##[debug]      a { color: #4183c4; text-decoration: none; }
##[debug]      a:hover { text-decoration: underline; }
##[debug]
##[debug]      h1 { letter-spacing: -1px; line-height: 60px; font-size: 60px; font-weight: 100; margin: 0px; text-shadow: 0 1px 0 #fff; }
##[debug]      p { color: rgba(0, 0, 0, 0.5); margin: 10px 0 10px; font-size: 18px; font-weight: 200; line-height: 1.6em;}
##[debug]
##[debug]      ul { list-style: none; margin: 25px 0; padding: 0; }
##[debug]      li { display: table-cell; font-weight: bold; width: 1%; }
##[debug]
##[debug]      .logo { display: inline-block; margin-top: 35px; }
##[debug]      .logo-img-2x { display: none; }
##[debug]      @media
##[debug]      only screen and (-webkit-min-device-pixel-ratio: 2),
##[debug]      only screen and (   min--moz-device-pixel-ratio: 2),
##[debug]      only screen and (     -o-min-device-pixel-ratio: 2/1),
##[debug]      only screen and (        min-device-pixel-ratio: 2),
##[debug]      only screen and (                min-resolution: 192dpi),
##[debug]      only screen and (                min-resolution: 2dppx) {
##[debug]        .logo-img-1x { display: none; }
##[debug]        .logo-img-2x { display: inline-block; }
##[debug]      }
##[debug]
##[debug]      #suggestions {
##[debug]        margin-top: 35px;
##[debug]        color: #ccc;
##[debug]      }
##[debug]      #suggestions a {
##[debug]        color: #666666;
##[debug]        font-weight: 200;
##[debug]        font-size: 14px;
##[debug]        margin: 0 10px;
##[debug]      }
##[debug]
##[debug]    </style>
##[debug]  </head>
##[debug]  <body>
##[debug]
##[debug]    <div class="container">
##[debug]      <p>
##[debug]        <img width="200" src="data:image/png;base64,
##[debug]iVBORw0KGgoAAAANSUhEUgAAAZAAAAGZCAMAAACQbpc2AAADAFBMVEWEBz6FAD6FAD6GAD+MAEGOAEOKAEGOAEOGAD+IAECOAEOOAUOOAEOOAEOOAEOOAEOOAEOOAEOCACyOAEQAAACKAEJpoJ2KADqu0eSKAD2BAD+KAD6AxCNqwoX1Ziawo9LYnGuwhL7aiwaIyYudst6DlrnhcWvCcVvlbE9PvsL2u3uYmdCZum9Rns7clUYsreOYp2nuj37kiZLUfqhvxafqfmrNWEjA22uz1Vu6lsiRAArWhLfgdoONnF36rWrzpz/ajy/LbqR8pdWDVXx8faN6m0+6hE3+1pmHP2mTy1b4k1p/ueFlj7vMfzyBs1v/2GfLpFL5rQD+7teQABj4n0/rpCz6r1Z8apP+4rrfw1iV0JyLUkqQACTDnah5AACtaEt9zcGNhleJbU5lsOFRvE5FuOPS4WLSt7+8ajvXcJpcyNvhztSwUkTGgZz9uDSxRnPZwsnTkqy7jpsPt/AxyvnAeJMAxPnMiqPJYIyzf4/KqrPedaK5bIsQwfTo2t72hUepPWvn6mO8VoXrgbDRaJXle6r75FzZm7S2TnyTNUXv5eiDAA6j0WmKyWv51liRAACjP0OZNVuwa4L+uQAcw/SiNGH57mIxxPKmU3DgpL2sX3rutc3/vh9gu1D3tNEAtvOSKVAmxfagRmbnrcWZJVaTJET2qsv84eyk04PzcEb71OX0u9OUzHD1nsR+1PhtwFUBvfQ5x/RizvdQy/VZzPWaz3b8uST+wFWDABqKAC6m03uGx2X+xGKBxmL2iWDQjb777/Nt0Pf+xmlCyPT+v032hVv+yXD3lGx2wlr99/hJyvX1gFV20vf1e1DCaqr6xdz+wlzJe7T2jmWFACT3i7v+0IP9u0R8xF7Oh7r+zHn/+WLHc7CQGUP3vtePEUT0i7qHADORIUyEAC2BAACf0Hr3j73zhLX7x96RGkrMgbePFEiRDUeOA0P4wNiVAEaPBkT
##[debug]      </p>
##[debug]
##[debug]      <p><strong>We had issues producing the response to your request.</strong></p>
##[debug]      <p>Sorry about that. Please try refreshing and contact us if the problem persists.</p>
##[debug]      <div id="suggestions">
##[debug]        <a href="https://github.com/contact">Contact Support</a> &mdash;
##[debug]        <a href="https://www.githubstatus.com/">GitHub Status</a> &mdash;
##[debug]        <a href="https://twitter.com/githubstatus">@githubstatus</a>
##[debug]      </div>
##[debug]
##[debug]      <a href="/" class="logo logo-img-1x">
##[debug]        <img width="32" height="32" title="" alt="" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyRpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNiAoTWFjaW50b3NoKSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDpFMTZCRDY3REIzRjAxMUUyQUQzREIxQzRENUFFNUM5NiIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDpFMTZCRDY3RUIzRjAxMUUyQUQzREIxQzRENUFFNUM5NiI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJl
##[debug]      </a>
##[debug]
##[debug]      <a href="/" class="logo logo-img-2x">
##[debug]        <img width="32" height="32" title="" alt="" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyRpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNiAoTWFjaW50b3NoKSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDpEQUM1QkUxRUI0MUMxMUUyQUQzREIxQzRENUFFNUM5NiIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDpEQUM1QkUxRkI0MUMxMUUyQUQzREIxQzRENUFFNUM5NiI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJl
##[debug]      </a>
##[debug]    </div>
##[debug]  </body>
##[debug]</html>
##[debug]
/home/runner/work/_actions/actions/download-artifact/eaceaf801fd36c7dee90939fad912460b18a1ffe/dist/index.js:2889
                throw new Error(`Failed to ${method}: ${error.message}`);
                      ^
Error: Failed to GetSignedArtifactURL: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
    at ArtifactHttpClient.<anonymous> (/home/runner/work/_actions/actions/download-artifact/eaceaf801fd36c7dee90939fad912460b18a1ffe/dist/index.js:2889:23)
    at Generator.throw (<anonymous>)
    at rejected (/home/runner/work/_actions/actions/download-artifact/eaceaf801fd36c7dee90939fad912460b18a1ffe/dist/index.js:2841:65)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Node.js v20.18.0
##[debug]Node Action run completed with exit code 1
##[debug]Finishing: Download Coverage Artifacts
```

I also found https://github.com/actions/toolkit/pull/1721 which was an attempt to improve the client's handling of 502. I'm not sure if this is a regression of that fix, or if it's a new issue. I don't see anything in the debug logs that indicate the request was retried.

**Comment by robherley** [comment]
👋 @contrast-jproberts 

You are using `@actions/download-artifact` at commit [`eaceaf801fd36c7dee90939fad912460b18a1ffe`](https://github.com/actions/download-artifact/blob/eaceaf801fd36c7dee90939fad912460b18a1ffe/), which is fairly old.

If you look at the package.json for that commit, it relies on an old version of `@actions/artifact` (`v2.1.1`):
- https://github.com/actions/download-artifact/blob/eaceaf801fd36c7dee90939fad912460b18a1ffe/package.json#L31

The tiny mitigation introduced in https://github.com/actions/toolkit/pull/1721 was released in `v2.1.6` of the `@actions/artifact` package. If you update to a more recent version of these actions, it should properly retry on these 502s in both upload/download artifact.

**Comment by contrast-jproberts** [comment]
😥 Thanks @robherley ! This project uses Dependabot and I assumed we were on the latest version of the action, but clearly that was wrong. I'll need to look into that next.

I'll upgrade to the latest version and close this issue since it's related to an older version. Thanks again for quickly pointing that out!
